### DISCUSSION TITLE
Is it possible to refer to all `$....Table` classes  in sort of clean, generic way?

### DISCUSSION THREAD

**Discussion by KarimElsayad247** [simolus3/drift#2283] [Q&A] [answered]
What I have are these three generated properties of the abstract `$_Database` class:

```dart
  late final $CommuteRoutesTable commuteRoutes = $CommuteRoutesTable(this);
  late final $CommuteRecordsTable commuteRecords = $CommuteRecordsTable(this);
  late final $RouteSegmentsTable routeSegments = $RouteSegmentsTable(this);
```

what I want is to have some class, let's call it `BaseController` where I can implement some common methods that I would want to implement for all my models. For now, `BaseController` looks like this:

```dart
abstract class BaseController extends ChangeNotifier {
  Database get db => getIt<Database>();
}
```

And, for example, I have a `CommuteRoutesController` which looks like this

```dart
class CommuteRoutesController extends BaseController {

  // ...bunch of stuff

  Future<List<CommuteRoute>> all() async {
    return db.select(db.commuteRoutes).get();
  }

 //... more bunch of stuff
```

What I want to do is move `all()` to the abstract class, closest I got is this implementation for `BaseController`

```dart
import 'package:drift/drift.dart' as d;
// ...other imports

abstract class BaseController<T extends d.Table, D> extends ChangeNotifier {
  Database get db => getIt<Database>();

  d.TableInfo<T, D> get table;

  Future<List<D>> all() async {
    return db.select(table).get();
  }
}
```

And this is `CommuteRoutesController`

```dart
class CommuteRoutesController
    extends BaseController<$CommuteRoutesTable, CommuteRoute> {
  @override
  TableInfo<$CommuteRoutesTable, CommuteRoute> get table => db.commuteRoutes;
// ...rest of class
```

and well.. **This works**, but, is there a better, cleaner way? or maybe some wheel I'm reinventing?

If you can tell, I'm seemingly trying to replicate ActiveRecord capabilities on top of drift :) In particular, I would like to keep code repetition to a minimum.

**Comment by simolus3** [comment, marked as answer]
> but, is there a better, cleaner way?

Is there anything in particular that you're unhappy with about those controllers? FWIW, these controllers don't even need to be abstract if you make `table` a final field and give them a constructor. Maybe I'm missing the point here, but having a single class for simple operations on all tables would be quite concise, right?

In drift, some utility methods are defined [as extensions on generated tables](https://github.com/simolus3/drift/blob/develop/drift/lib/src/runtime/query_builder/on_table.dart), does that go into rough direction you were aiming for? I think something like `all()` would be a reasonable extension for that as well, I'll add it.

So in general, your approach of `BaseController`s looks reasonable to me. If you have ideas for changes in drift that would make them easier to write, that would be worth discussing. Also, if you have a suggestion for queries that should already be available in drift as a go-to method, those ideas would be welcome as well.

**Reply by KarimElsayad247** [reply]
Hi simolus, sorry for the (very) late reply, I got distracted with some things :)

What rang an alram in my head is my feeling that I had to dig through library and generated code to find something that would represent a common interface for all tables. (the `TableInfo<T, D>`).

Indeed the class is quite concise, but I don't know if I'm using something stable, or some change might completely break it because I'm relying on an undocumented implementation detail. Of course, it was just now that I [realized there's some api docs](https://drift.simonbinder.eu/api/drift/tableinfo-mixin)

I apologize if I was being confusing, perhaps that stems from being new to dart and flutter :) I wasn't aware of Dart extensions, they same like just the right tool for the job. I'll try looking into them more.

BTW I saw that `all()` was added! that's cool!

**Reply by simolus3** [reply]
Basically, everything in drift that isn't marked as `@internal` or explicitly called experimental is part of the public API, you can expect that to stay stable.

> What rang an alram in my head is my feeling that I had to dig through library and generated code to find something that would represent a common interface for all tables

That's a fair point - most users don't need these reflection-like capabilities over their database schema, but it is a powerful tool that should perhaps be mentioned in the documentation somewhere.
